packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -959,3 +921,1328 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
+      assert(oldWidget.entry == widget.entry);
+      theatre = context.findAncestorRenderObjectOfType<_RenderTheatre>()!;
+      _cachedOverlayEntryTopChildModel?._dispose();
+      _cachedOverlayEntryTopChildModel = null;
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.entry._overlayStateMounted.value = null;
+    widget.entry._didUnmount();
+    _cachedOverlayEntryTopChildModel?._dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return TickerMode(
+      enabled: widget.tickerEnabled,
+      child: _RenderTheatreMarker(
+        theatre: theatre,
+        overlayEntryWidgetState: this,
+        child: widget.entry.builder(context),
+      ),
+    );
+  }
+
+  void _markNeedsBuild() {
+    setState(() { /* the state that changed is in the builder */ });
+  }
+}
+
+
+// * OverlayPortal Implementation
+//  OverlayPortal is inspired by the
+//  [flutter_portal](https://pub.dev/packages/flutter_portal) package.
+//
+// ** RenderObject hierarchy
+// The widget works by inserting its overlay child's render subtree directly
+// under [Overlay]'s render object (_RenderTheatre).
+// https://user-images.githubusercontent.com/31859944/171971838-62ed3975-4b5d-4733-a9c9-f79e263b8fcc.jpg
+//
+// To ensure the overlay child render subtree does not do layout twice, the
+// subtree must only perform layout after both its _RenderTheatre and the
+// [OverlayPortal]'s render object (_RenderLayoutSurrogateProxyBox) have
+// finished layout. This is handled by _RenderDeferredLayoutBox.
+//
+// ** Z-Index of an overlay child
+// [_OverlayLocation] is a (currently private) interface that allows an
+// [OverlayPortal] to insert its overlay child into a specific [Overlay], as
+// well as specifying the paint order between the overlay child and other
+// children of the _RenderTheatre.
+//
+// Since [OverlayPortal] is only allowed to target ancestor [Overlay]s
+// (_RenderTheatre must finish doing layout before _RenderDeferredLayoutBox),
+// the _RenderTheatre should typically be acquired using an [InheritedWidget]
+// (currently, _RenderTheatreMarker) in case the [OverlayPortal] gets
+// reparented.
+
+/// A class to show, hide and bring to top an [OverlayPortal]'s overlay child
+/// in the target [Overlay].
+///
+/// A [OverlayPortalController] can only be given to at most one [OverlayPortal]
+/// at a time. When an [OverlayPortalController] is moved from one
+/// [OverlayPortal] to another, its [isShowing] state does not carry over.
+///
+/// The [show] and [hide] methods typically should not be called when the widget
+/// tree is being rebuilt. However it's allowed to call [show] or [hide] when
+/// the associated [OverlayPortal] widget is yet to be inflated into the widget
+/// tree.
+class OverlayPortalController {
+  /// Creates an [OverlayPortalController].
+  OverlayPortalController({ String? debugLabel }) : _debugLabel = debugLabel;
+
+  _OverlayPortalState? _attachTarget;
+  int? _showTimestamp;
```

Comment:
It's not easy to maintain temporary data structures that contain reference to OverlayPortals in Overlay, since an OverlayPortal can be reparented to a different Overlay in the next frame.

Reply:
the idea i was proposing is a temporary data structure that will be flushed during the draw frame, so it will be empty at the end of frame.

My thinking was, if the purpose of _showTimestamp is to keep track of the order of the OverlayPortalController.show() on the OverlayEntry in the same frame. Why not just store the OverlayEntry in a list by the order of OverlayPortalController.show() calls.

Reply:
When an `OverlayPortal` gets reparented to a different `Overlay`, `_showTimestamp` would be used to determine the new location of the overlay child in the new `Overlay`'s child model.

Reply:
I guess this is more like a design question, and my opinion may be bias. Are we sure we should preserve the order when reparenting? One possible scenario is a widget with global key that gets build in a different place. To the developer, the widget in the new location is 'newly' build and the overlay may want to be at the top? Another question is if we do have a declarative z-index API as you mentioned in https://github.com/flutter/flutter/pull/105335#discussion_r963284676. Maybe we should implement that directly and solve the ordering problem. WDYT?

Reply:
Do you think this would be a problem if the system time change somehow?